api: Support assigning qualified qmlelem to a property

stephenmdangelo opened the pull request:
e.g.

import QtQuick 2.5 as QQ

Repeater {
  delegate: QQ.Rectangle {}
}

ChALkeR commented:
Sorry for not looking at this earlier.

This shouldn't change the existing testcase (removing the test that we already have, i.e. `property Item item: Item {}`). Instead, it should introduce a new one.

In this specific case either _adding_ a new `import` line and adding some new properties to `Properties.qml` or adding an entirely new file would work.

ChALkeR commented:
Also I think that this comment by @akreuzkamp: https://github.com/qmlweb/qmlweb-parser/pull/17#issuecomment-231526384 is also applicable to this PR.
Shouldn't the names be converted to proper AST? For example, see [/ImportsNamed.qml](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml) and [ImportsNamed.qml.json](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml.json) files.

ChALkeR commented on tests/qml/Properties.qml.json:
```diff
@@ -5,13 +5,13 @@
   [
     "qmlelem",
-    "Rectangle",
+    "QQ.Rectangle",
```
This is not the proper format for `qmlelem`. See [ImportsNamed.qml.json#L27-L32](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml.json#L27-L32) for how qualified `qmlelem` should be encoded. You could also test with longer paths. e.g. `Abc.Def.Rectangle`.

ChALkeR replied:
This has changed and will be `["dot",  "QQ", "Rectangle"],` after a rebase.
Note that this line shouldn't even be present here, as `Rectangle` → `QQ.Rectangle` is an unrelated change on this line.

ChALkeR commented on tests/qml/Properties.qml.json:
```diff
@@ -131,12 +131,12 @@
           [
             "qmlelem",
-            "Item",
+            "QQ.Item",
```
This is not the proper format for `qmlelem`. See [ImportsNamed.qml.json#L27-L32](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml.json#L27-L32) for how qualified `qmlelem` should be encoded. You could also test with longer paths. e.g. `Abc.Def.Rectangle`.

ChALkeR commented on tests/qml/Properties.qml:
```diff
@@ -12,5 +12,5 @@ Rectangle {
   property var foo: {}
   property var bar: []
-  property Item item: Item {}
+  property Item item: QQ.Item {}
```
This should be added as a new line, and should not replace the current `property Item item: Item {}` test.

ChALkeR commented on tests/qml/Properties.qml:
```diff
@@ -1,6 +1,6 @@
+import QtQuick 2.5 as QQ
 
-Rectangle {
+QQ.Rectangle {
```
Unneeded change.

ChALkeR commented on tests/qml/Properties.qml:
```diff
@@ -1,6 +1,6 @@
-import QtQuick 2.5
+import QtQuick 2.5 as QQ
```
The original line should be present to keep the existing testcase (see comment below).
Perhaps use some other import, not `QtQuick`, and add it to the second line?

ChALkeR commented:
@stephenmdangelo, sorry for not reviewing at this earlier. In general, I'm in favor of this PR (and thanks for doing this!), but it requires some changes before landing. See the review above.

stephenmdangelo commented:
@ChALkeR So I've cleaned up the Properties.json test (as well, I added a change to `tape.js` to fix the tests failing if, say, I was editing in vi).

I think I'm generating the dot expression as you're suggesting.

Though I find it odd that the `expr` as returned by `maybe_assign` is of the form:

```
[
  "dot",
  "DOM",
  [
    "name",
    "DOM"
  ],
  "Div"
]
```
But a non-assigned `qmlelem` (as in `ImportsNamed.qml`) omits the `name` array:

```
[
   "dot",
   "DOM",
   "Div"
]
```

So I went with the latter form. Is this correct? If so, why don't we use the `name` format, or why does that format get returned by `maybe_assign`?

ChALkeR commented on tests/tape.js:
```diff
@@ -14,6 +14,9 @@ function buildTree(dir, data) {
   for (const file of fs.readdirSync(dir)) {
     const filePath = path.join(dir, file);
+    if (file[0] === ".") {
+        continue;
```
Nit: should be above `const filePath`.

ChALkeR commented:
I will take a look at `maybe_assign` a bit later, though. =)
